quickwit-storage/src/bundle_storage.rs

```diff
@@ -0,0 +1,282 @@
+        self.storage
+            .get_slice(&self.bundle_file_name, new_range)
+            .await
+    }
+
+    async fn get_all(&self, path: &Path) -> crate::StorageResult<Bytes> {
+        let file_offsets = self.metadata.get(path)?;
+        self.storage
+            .get_slice(
+                &self.bundle_file_name,
+                file_offsets.start as usize..file_offsets.end as usize,
+            )
+            .await
+    }
+
+    async fn delete(&self, path: &Path) -> crate::StorageResult<()> {
+        Err(unsupported_operation(path))
+    }
+
+    async fn exists(&self, path: &Path) -> crate::StorageResult<bool> {
+        // also check if self.bundle_file_name exists ?
+        Ok(self.metadata.exists(path))
+    }
+
+    fn uri(&self) -> String {
+        self.storage.uri()
+    }
+}
+
+#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
+struct FileOffsets {
+    start: u64,
+    end: u64,
+}
+
+impl HasLen for BundleStorage {
+    fn len(&self) -> usize {
+        unimplemented!()
+    }
+}
+
+impl fmt::Debug for BundleStorage {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "BundleStorage({:?}, files={:?})",
+            &self.bundle_file_name, self.metadata
+        )
+    }
+}
+
+fn unsupported_operation(path: &Path) -> StorageError {
+    let msg = "Unsupported operation. BundleStorage only supports async reads";
+    error!(path=?path, msg);
+    io::Error::new(io::ErrorKind::Other, format!("{}: {:?}", msg, path)).into()
+}
+
+/// BundleStorage bundles together multiple files into a single file
+/// with some metadata
+pub struct BundleStorageBuilder {
+    metadata: BundleStorageMetadata,
+    current_offset: u64,
+    /// The offset in the file where the hotcache begins. This is used to read
+    /// the hotcache and footer in a single read.
+    pub hotcache_offset: u64,
+    /// The footer offset, where the footer begins.
+    /// The footer includes the footer metadata as json encoded and the size of the footer in
+    /// bytes.
+    pub footer_offset: u64,
+    bundle_file: File,
+}
+
+impl BundleStorageBuilder {
+    /// Creates a new CreateBundleStorage, to which files can be added.
+    pub fn new(path: &Path) -> io::Result<Self> {
+        let sink = OpenOptions::new().create(true).append(true).open(path)?;
+
+        Ok(BundleStorageBuilder {
+            bundle_file: sink,
+            current_offset: 0,
+            hotcache_offset: 0,
+            footer_offset: 0,
+            metadata: Default::default(),
+        })
+    }
+
+    /// Appends a file to the bundle file.
+    ///
+    /// The hotcache needs to be the last file that is added, in order to be able to read
+    /// the hotcache and the metadata in one continous read.
+    pub fn add_file(&mut self, path: &Path) -> io::Result<()> {
+        let mut file = File::open(path)?;
+        let bytes_written = io::copy(&mut file, &mut self.bundle_file)?;
```

Comment:
Nitpick. 

if you want to make your life easier for unit test, you can add another public function that takes a 

```
fn add_file_from_read(&Path, read: &mut io::Read) -> io::Result<()>;
```

And call it from your `add_file` method.

Reply:
(I let you decide)